### PR TITLE
HOSTEDCP-1429: hypershift-operator: add a controller for t-shirt sizing

### PR DESCRIPTION
**Pull request by stevekuznetsov**
/assign @csrwng

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
```
// TODO: pause controller when reconciliation is paused
```

**Comment by stevekuznetsov** [comment]
@csrwng integration test:

```
$ GO_TEST_FLAGS='-run TestHostedSizingController' PATH="${PATH}:$(realpath ./bin)" WORK_DIR=/tmp/integration PULL_SECRET=~/secrets/ocm-pull-secret ./test/integration/run.sh test
Running test...
2024-03-06T13:13:07-07:00	INFO	running tests
=== RUN   TestHostedSizingController
    hosted_cluster_sizing_test.go:33: setting the cluster sizing configuration
    hosted_cluster_sizing_test.go:78: adding a small node count (1) to the hosted control plane hosted-clusters-2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8
    hosted_cluster_sizing_test.go:184: expecting hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to transition to small size
    hosted_cluster_sizing_test.go:210: waiting for hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to have status condition ClusterSizeComputed=True: small and match validation func
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2147 after 9.721094ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=True: TransitionDelayNotElapsed (HostedClusters must wait at least 1h0m0s to increase in size after the last transition.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: small (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:254: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 validation: hostedCluster.metadata.labels[io.openshift.hypershift/hosted-cluster-size]=small
    hosted_cluster_sizing_test.go:91: adding a large node count (50) to the hosted control plane hosted-clusters-2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8
    hosted_cluster_sizing_test.go:101: expecting hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to delay transition
    hosted_cluster_sizing_test.go:210: waiting for hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to have status condition ClusterSizeTransitionPending=True: TransitionDelayNotElapsed
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2149 after 9.176652ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=True: TransitionDelayNotElapsed (HostedClusters must wait at least 1h0m0s to increase in size after the last transition.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: small (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:107: configure the cluster sizing configuration to have miniscule delays
    hosted_cluster_sizing_test.go:184: expecting hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to transition to medium size
    hosted_cluster_sizing_test.go:210: waiting for hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to have status condition ClusterSizeComputed=True: medium and match validation func
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2149 after 2.248304ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=True: TransitionDelayNotElapsed (HostedClusters must wait at least 1h0m0s to increase in size after the last transition.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: small (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2153 after 116.688634ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: medium (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=False: ClusterSizeTransitioned (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:254: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 validation: hostedCluster.metadata.labels[io.openshift.hypershift/hosted-cluster-size]=medium
    hosted_cluster_sizing_test.go:128: configure the cluster sizing configuration to have no concurrency
    hosted_cluster_sizing_test.go:148: adding a huge node count (500) to the hosted control plane hosted-clusters-2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8
    hosted_cluster_sizing_test.go:158: expecting hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to delay transition for concurrency
    hosted_cluster_sizing_test.go:210: waiting for hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to have status condition ClusterSizeTransitionPending=True: ConcurrencyLimitReached
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2153 after 2.245286ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: medium (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=False: ClusterSizeTransitioned (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2157 after 100.333364ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=True: TransitionDelayNotElapsed (HostedClusters must wait at least 1s to increase in size after the last transition.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: medium (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2159 after 513.554479ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=True: ConcurrencyLimitReached (1 HostedClusters have already transitioned sizes in the last 1h0m0s, more time must elapse before the next transition.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: medium (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:164: configure the cluster sizing configuration to have allow lots of concurrency
    hosted_cluster_sizing_test.go:184: expecting hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to transition to large size
    hosted_cluster_sizing_test.go:210: waiting for hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 to have status condition ClusterSizeComputed=True: large and match validation func
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2159 after 2.33001ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=True: ConcurrencyLimitReached (1 HostedClusters have already transitioned sizes in the last 1h0m0s, more time must elapse before the next transition.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: medium (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:224: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 observed at RV 2163 after 104.618861ms
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeComputed=True: large (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:236: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 status: ClusterSizeTransitionPending=False: ClusterSizeTransitioned (The HostedCluster has transitioned to a new t-shirt size.)
    hosted_cluster_sizing_test.go:254: hosted cluster hosted-clusters/2sbjc6akxi6ochgq0u1y0u6gah0ica9ph1s668m92fw8 validation: hostedCluster.metadata.labels[io.openshift.hypershift/hosted-cluster-size]=large
    run.go:79: cleaning up...
--- PASS: TestHostedSizingController (0.98s)
PASS
ok  	github.com/openshift/hypershift/test/integration	0.998s
```

File: cmd/install/install.go [hunk 1]
```diff
@@ -206,6 +209,7 @@ func NewCommand() *cobra.Command {
 	cmd.PersistentFlags().BoolVar(&opts.EnableUWMTelemetryRemoteWrite, "enable-uwm-telemetry-remote-write", opts.EnableUWMTelemetryRemoteWrite, "If true, HyperShift operator ensures user workload monitoring is enabled and that it is configured to remote write telemetry metrics from control planes")
 	cmd.PersistentFlags().BoolVar(&opts.EnableCVOManagementClusterMetricsAccess, "enable-cvo-management-cluster-metrics-access", opts.EnableCVOManagementClusterMetricsAccess, "If true, the hosted CVO will have access to the management cluster metrics server to evaluate conditional updates (supported for OpenShift management clusters)")
 	cmd.Flags().BoolVar(&opts.WaitUntilAvailable, "wait-until-available", opts.WaitUntilAvailable, "If true, pauses installation until hypershift operator has been rolled out and its webhook service is available (if installing the webhook)")
+	cmd.Flags().BoolVar(&opts.WaitUntilEstablished, "wait-until-established", opts.WaitUntilEstablished, "If true, pauses installation until all custom resource definitions are established before applying other manifests.")
```
Comment (csrwng) [inline cmd/install/install.go:212]: nit: wait-until-established doesn't tell me much about what it's about.
How about `wait-until-crds-established` ?

**Review by csrwng — COMMENTED** [review]
nits

File: hypershift-operator/controllers/hostedclustersizing/hostedclustersizing_controller.go [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+	metav1applyconfigurations "k8s.io/client-go/applyconfigurations/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	hypershiftclient "github.com/openshift/hypershift/client/clientset/clientset"
+)
+
+const (
+	HostedClusterSizeLabel = "io.openshift.hypershift/hosted-cluster-size"
```
Comment (csrwng) [inline hypershift-operator/controllers/hostedclustersizing/hostedclustersizing_controller.go:28]: pls change to `hypershift.openshift.io/hosted-cluster-size`

**Comment by csrwng** [comment]
/hold
@stevekuznetsov in manually testing this, I don't see the delay taking effect on scale up/down. I changed the config after the hypershift operator was already running, so not sure if the config is stale or it's something else.

**Comment by stevekuznetsov** [comment]
```
 Error: failed to create infra: cannot tag NAT gateway EIP: InvalidElasticIpID.NotFound: The elasticIp ID 'eipalloc-02b301c4c6f314eec' does not exist
	status code: 400, request id: c6d39334-9eac-4b65-a468-1971e77a3cfc
failed to create infra: cannot tag NAT gateway EIP: InvalidElasticIpID.NotFound: The elasticIp ID 'eipalloc-02b301c4c6f314eec' does not exist
	status code: 400, request id: c6d39334-9eac-4b65-a468-1971e77a3cfc 
```

/retest

**Comment by csrwng** [comment]
@stevekuznetsov one thing that occurred to me is that we should skip reconcile of a hostedcluster that either has a deletionTimestamp or is [paused](https://github.com/openshift/hypershift/blob/ae1caa00ff3a2c2bfc1129f0168efc1e786d1d12/api/hypershift/v1beta1/hostedcluster_types.go#L447-L452). Other than that, lgtm

File: hypershift-operator/controllers/hostedclustersizing/hostedclustersizing_controller.go [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+			break
+		}
+	}
+	if !configValid {
+		// we can't put clusters into t-shirt sizes unless we have a valid configuration; we'll re-trigger when
+		// the configuration object changes and can process clusters then
+		return nil, nil
+	}
+
+	if hostedCluster.ObjectMeta.DeletionTimestamp != nil {
```
Comment (csrwng) [inline hypershift-operator/controllers/hostedclustersizing/hostedclustersizing_controller.go:179]: really minor nit:
s/hostedCluster.ObjectMeta.DeletionTimestamp != nil/hostedCluster.DeletionTimestamp.IsZero()/

**Pull request merged**